### PR TITLE
pull in user name and email from university-web instead of github

### PR DESCRIPTION
**Pull request by ericgj**
For issue #6. 

I only have tested it in development, where it runs without errors, but not really adequate since the UniversityWeb::User service short-circuits when in development.  (For some reason having trouble setting up the test environment but I'll keep trying.)

Anyway, see what you think of the code anyway. I'm not that crazy about it, but maybe looking at it other approaches will come to mind.

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -46,9 +46,20 @@ def auth_hash
...
     nick = auth_hash['info']['nickname']
-    user = UniversityWeb::User.find_by_github(nick)
+    @uniweb_user ||= UniversityWeb::User.find_by_github(nick)
+  end
+  
+  def user_hash
+    auth_hash.merge(:name => uniweb_user.name, 
+                    :email => uniweb_user.email
+                   )
+  end
```
Comment (ericgj) [inline app/controllers/sessions_controller.rb:60]: The main thing I don't like so much is that this `user_hash` (merging the github with the uni-web user info) depends upon the before filter having run, and redirected if uni-web user doesn't exist or doesn't have permissions -- i.e. currently must be called in the `create` and nowhere else. So maybe should just get rid of this separate method and do it directly in the `create`.

**Comment by jordanbyron** [comment]
> (For some reason having trouble setting up the test environment but I'll keep trying.)

Try pulling from upstream. I pushed a fix for the tests not too long ago. That might solve some of your problems. It also looks like `SessionsController#check_permissions` has some bugs in it. Old references to `nick` which is no longer present.

> I'm not that crazy about it, but maybe looking at it other approaches will come to mind.

It's not terrible, but I'm also becoming concerned with the increasing amount of code in that controller. However @sandal and I are starting to throw around the idea of a "Mendicant Identity" service which would replace all of the University-Web checks in Community.

With that in mind, this is probably "OK for now". See if you can't get the tests green and we'll go from there :smile:

**Comment by ericgj** [comment]
I fixed it up a bit. Test pass (now that I finally fixed my postgres environment). 

But this feature isn't being covered by tests at all - and hard to say how to do that since it involves two services. 

PS. sorry about the weird timestamps on my commits ... at some point I have to get to the bottom of why my virtualbox cannot keep time!

**Comment by jordanbyron** [comment]
> But this feature isn't being covered by tests at all - and hard to say how to do that since it involves two services.

I think this would be really easy to do using the [auth helpers](https://github.com/mendicant-university/community/blob/master/test/support/auth.rb) I'm pretty sure you wrote for stream. I imagine it would be an integration test that would use `mock_auth_for` and `mock_uniweb_user` checking for the various types of scenarios:
- User not in university web
- User not alumni or staff
- User alumni or staff
- Name / Email used from university web when creating user

Does that sound like something you could bang out?

> I fixed it up a bit.

Thanks. Looks better to me :smile:

**Comment by ericgj** [comment]
I keep thinking, "this seems familiar"   :doh:  forgot I did that.  ok, sure maybe tomorrow.

**Comment by jordanbyron** [comment]
Sweet. I'll hold off merging this until we've got those tests in place. Thanks for working on this! :trophy:

**Comment by ericgj** [comment]
I've got all these test passing except not sure what to do with this one:

> Name / Email used from university web when creating user

Do you want me to mock out User.create_from_hash with mocha?

**Comment by jordanbyron** [comment]
> Do you want me to mock out User.create_from_hash with mocha?

Nope. Should be able to do something like:

``` ruby
OmniAuth.config.mock_auth[:github] = {
  'provider' => 'github',
  'uid'      => '1234',
  'info'     => {
    'name'     => 'franky',
    'nickname' => 'frankpepelio',
    'email'    => nil
  }
}

mock_uniweb_user(
  'github'           => 'frankpepelio',
  'alumnus'          => true,
  'staff'            => false,
  'visiting_teacher' => false,
  'name'             => 'Frank Pepelio',
  'email'            => 'frank.pepelio@gmail.com'
)

visit 'auth/github'

user = User.find_by_github('frankpepelio')

assert_equal 'Frank Pepelio', user.name
assert_equal 'frank.pepelio@gmail.com', user.email
```

**Comment by ericgj** [comment]
Last test still failing. Might be due to some kind of leftover session hanging around between tests. Do I remember some setting Capybara has for dealing with this?

EDIT: I see we already have Capybara.reset_sessions!  in the teardown. So I don't know WTF is going on.  Here's the relevant logs.
Clearly, somehow in going to the root_path, a new user gets created, but not through the SessionsController#create.  Then when you go to /auth/github, the user is already created, so it doesn't have a chance to merge in the uni-web info.

```
Started GET "/" for 127.0.0.1 at 2012-02-16 18:36:06 -0500
  Processing by ArticlesController#index as HTML
Rendered articles/_header.html.haml (1.9ms)
  [1m[36mArticle Load (0.3ms)[0m  [1mSELECT "articles".* FROM "articles" ORDER BY created_at desc LIMIT 30 OFFSET 0[0m
Rendered articles/_article.html.haml (0.0ms)
Rendered shared/_flash.html.haml (0.1ms)
  [1m[35mUser Load (0.3ms)[0m  SELECT "users".* FROM "users" WHERE "users"."id" = 2 LIMIT 1
Rendered shared/_navigation.html.haml (4.0ms)
Rendered shared/_section_header.html.haml (0.1ms)
Rendered shared/_footer.html.haml (0.5ms)
Completed 200 OK in 15ms (Views: 12.5ms | ActiveRecord: 0.6ms)
  [1m[36m (0.1ms)[0m  [1mROLLBACK[0m
  [1m[35m (0.1ms)[0m  BEGIN
  [1m[36m (0.1ms)[0m  [1mSAVEPOINT active_record_1[0m
  [1m[35m (0.2ms)[0m  SELECT 1 FROM "users" WHERE "users"."github" = 'bartsimpson' LIMIT 1
  [1m[36mSQL (0.4ms)[0m  [1mINSERT INTO "users" ("admin", "created_at", "description", "email", "github", "github_id", "name", "twitter", "uid", "updated_at", "website") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11) RETURNING "id"[0m  [["admin", false], ["created_at", Thu, 16 Feb 2012 23:36:06 UTC +00:00], ["description", nil], ["email", "faker@mcfakerstein.com"], ["github", "bartsimpson"], ["github_id", nil], ["name", "Bart Simpson"], ["twitter", nil], ["uid", "789"], ["updated_at", Thu, 16 Feb 2012 23:36:06 UTC +00:00], ["website", nil]]
  [1m[35m (0.1ms)[0m  RELEASE SAVEPOINT active_record_1


Started GET "/" for 127.0.0.1 at 2012-02-16 18:36:06 -0500
  Processing by ArticlesController#index as HTML
Rendered articles/_header.html.haml (1.0ms)
  [1m[36mArticle Load (0.3ms)[0m  [1mSELECT "articles".* FROM "articles" ORDER BY created_at desc LIMIT 30 OFFSET 0[0m
Rendered articles/_article.html.haml (0.0ms)
Rendered shared/_flash.html.haml (0.1ms)
  [1m[35mUser Load (0.3ms)[0m  SELECT "users".* FROM "users" WHERE "users"."id" IS NULL LIMIT 1
  [1m[36mCACHE (0.0ms)[0m  [1mSELECT "users".* FROM "users" WHERE "users"."id" IS NULL LIMIT 1[0m
Rendered shared/_navigation.html.haml (3.7ms)
  [1m[35mCACHE (0.0ms)[0m  SELECT "users".* FROM "users" WHERE "users"."id" IS NULL LIMIT 1
Rendered shared/_section_header.html.haml (0.5ms)
Rendered shared/_footer.html.haml (0.5ms)
Completed 200 OK in 13ms (Views: 11.6ms | ActiveRecord: 0.6ms)


Started GET "/login" for 127.0.0.1 at 2012-02-16 18:36:06 -0500
  Processing by SessionsController#new as HTML
Redirected to http://www.example.com/auth/github
Completed 302 Found in 1ms


Started GET "/auth/github" for 127.0.0.1 at 2012-02-16 18:36:06 -0500


Started GET "/auth/github/callback" for 127.0.0.1 at 2012-02-16 18:36:06 -0500
  Processing by SessionsController#create as HTML
  Parameters: {"provider"=>"github"}
[SessionsController#create] github auth: {"provider"=>"github", "uid"=>"789", "info"=>{"name"=>"Bart Simpson", "nickname"=>"bartsimpson", "email"=>"faker@mcfakerstein.com"}}
[SessionsController#create] uniweb_user: #<struct UniversityWeb::User github="bartsimpson", name="Evil Twin of Bart Simpson", email="dont.have.a.cow.man@gmail.com", alumnus=true, staff=false, visiting_teacher=false>
  [1m[36mUser Load (0.3ms)[0m  [1mSELECT "users".* FROM "users" WHERE "users"."uid" = '789' LIMIT 1[0m
Redirected to http://www.example.com/
Completed 302 Found in 2ms
```

**Comment by ericgj** [comment]
Thanks, knew it was something obvious. See what you think. It's still pretty ugly, but passes. 

Might be better to set up some specific github/uni-web fixtures rather than piggybacking off of the user factories, I think that's part of what's causing it to seem unnecessarily complicated.

**Comment by jordanbyron** [comment]
> It's still pretty ugly, but passes.

I DRYed up the auth support file a bit but overall this is exactly what it needs to be. Sooner than later we'll be replacing most of this with our new identity server, but this is going to make sure any new community members get the correct information. Thanks Eric! :moneybag:

**Pull request merged**